### PR TITLE
Feat: add stream output to exec_run

### PR DESCRIPTION
**Pull request by iFurySt**
* Using command timeout to control the exec_box's timeout.

https://github.com/OpenDevin/OpenDevin/assets/16201837/cf3e6276-b1bf-44b8-9333-57f824a650de

i found the init is too long and haven't feedback, and IMO i think the init does not even need to control the timeout. i failed to init a few times yesterday because my network seems poor, it made me a bit confused ~

i added the stream output for feedback and kept the timeout but changed to use the `timeout` command.

### CODE REVIEW

**Comment by rbren** [comment]
This is very cool!

I'm a little worried about changing the footprint of `execute`, which hypothetically should be the same for every box. But I don't think it affects much here.

Do you think it'd be possible to add this to SSH box too?

And do you think it'd be possible to use this to stream output to the FE?

**Comment by iFurySt** [comment]
> I'm a little worried about changing the footprint of execute, which hypothetically should be the same for every box.

i think every box should have a stream output as much as possible, they can easily implement the `CancellableStream`. (but i'm not so sure that will be the case for all in the future🤯)

> Do you think it'd be possible to add this to the SSH box too?

yes, i wanna apply the stream to the SSH box. but i was too busy the last few days, i think i can work at it by the end of the week.

> And do you think it'd be possible to use this to stream output to the FE?

i think so, but it may have a bit of trouble when the concurrent sends messages from the BE to FE.

**Comment by iFurySt** [comment]
https://github.com/OpenDevin/OpenDevin/assets/16201837/dad8b2f3-6747-4618-bb2e-9e9a7192c7cc

ignores the error for BrowserEnv, it has nothing to do with this pr
```
TypeError: BrowserEnv.__init__() got an unexpected keyword argument 'start_url' was raised from the environment creator for browsergym/openended with kwargs ({'start_url': 'about:blank', 'wait_for_user_message': False, 'headless': True})
```

**Comment by iFurySt** [comment]
@rbren @xingyaoww feel free to review it whenever you have time.

**Comment by rbren** [comment]
This is looking good to me!

I'm mostly worried about the complicated logic for parsing the SSH stream--nothing we can leverage from pexpect or another SSH lib?

**Comment by iFurySt** [comment]
> I'm mostly worried about the complicated logic for parsing the SSH stream--nothing we can leverage from pexpect or another SSH lib?

i can't found related lib to use. [paramiko](https://github.com/paramiko/paramiko) with [paramiko-expect](https://github.com/fgimian/paramiko-expect) are similar with expect. the parse logic just buffer the output and then return line by line, and remove the PROMPT([PEXPECT]) finally.

**Review by rbren — APPROVED** [review]

File: opendevin/runtime/docker/ssh_box.py [hunk 1]
```diff
@@ -344,21 +412,25 @@ def _send_interrupt(
             f'Command: "{cmd}" timed out. Sending SIGINT to the process: {command_output}',
         )
 
-    def execute(self, cmd: str, timeout: int | None = None) -> tuple[int, str]:
+    def execute(
+        self, cmd: str, stream: bool = False, timeout: int | None = None
```
Comment (SmartManoj) [inline opendevin/runtime/docker/ssh_box.py:416]: Why stream is disabled by default?

Comment (iFurySt) [reply opendevin/runtime/docker/ssh_box.py:416]: I think streaming output makes sense for initialization or time-consuming operations to get feedback. Currently, we don't support streaming output to the frontend. And the agent prefers not to receive the streaming output.

Comment (SmartManoj) [reply opendevin/runtime/docker/ssh_box.py:416]: How about only displaying the output to the logs?

**Comment by SmartManoj** [comment]
> related lib

@iFurySt [Paramiko with continuous stdout](https://stackoverflow.com/a/33334998) You were looking for this?

**Pull request merged**